app/com/arpnetworking/metrics/portal/reports/ReportExecution.java

```diff
@@ -0,0 +1,184 @@
+import models.internal.impl.HtmlReportFormat;
+import models.internal.impl.PdfReportFormat;
+import models.internal.reports.Recipient;
+import models.internal.reports.Report;
+import models.internal.reports.ReportFormat;
+import models.internal.reports.ReportSource;
+
+import java.time.Instant;
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Collectors;
+
+/**
+ * Utilities for execution of {@link Report}s.
+ *
+ * @author Spencer Pearson (spencerpearson at dropbox dot com)
+ */
+public final class ReportExecution {
+
+    /**
+     * Render and send a report.
+     *
+     * @param report The report to render and execute.
+     * @param injector A Guice injector to pull dependencies out of.
+     * @param scheduled The instant the report was scheduled for.
+     * @return A CompletionStage that completes when sending has completed for every recipient.
+     */
+    public static CompletionStage<Report.Result> execute(final Report report, final Injector injector, final Instant scheduled) {
+        verifyDependencies(report, injector);
+        final ImmutableMultimap<ReportFormat, Recipient> formatToRecipients = report.getRecipientsByFormat()
+                .entrySet()
+                .stream()
+                .collect(ImmutableSetMultimap.flatteningToImmutableSetMultimap(
+                        Map.Entry::getKey,
+                        e -> e.getValue().stream()));
+        final ImmutableMultimap<Recipient, ReportFormat> recipientToFormats = formatToRecipients.inverse();
+        return renderAll(injector, formatToRecipients.keySet(), report.getSource(), scheduled)
+                .thenCompose(
+                        formatToRendered -> sendAll(injector, recipientToFormats, formatToRendered)
+                ).thenApply(
+                        nothing -> null
+                );
+    }
+
+    /* package private */ static void verifyDependencies(final Report report, final Injector injector) {
+        for (final ReportFormat format : report.getRecipientsByFormat().keySet()) {
+            getRenderer(injector, report.getSource(), format);
+        }
+        final Collection<Recipient> allRecipients = report.getRecipientsByFormat().values().stream()
+                .flatMap(Collection::stream)
+                .collect(Collectors.toSet());
+        for (final Recipient recipient : allRecipients) {
+            getSender(injector, recipient);
+        }
+    }
+
+    /* package private */ static CompletionStage<ImmutableMap<ReportFormat, RenderedReport>> renderAll(
+            final Injector injector,
+            final ImmutableSet<ReportFormat> formats,
+            final ReportSource source,
+            final Instant scheduled
+    ) {
+        final Map<ReportFormat, RenderedReport> result = Maps.newHashMap(); // TODO(spencerpearson) -- ConcurrentMap?
```

Comment:
Yes. You are populating the map concurrently as the result of each render future is executed. Nothing guarantees synchronous access to the map.